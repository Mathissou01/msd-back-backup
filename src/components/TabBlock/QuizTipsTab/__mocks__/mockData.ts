import { GetQuizAndTipsBlockDocument } from "../../../../graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    result: {
      data: {
        contractCustomizations: {
          data: [
            {
              attributes: {
                homepage: {
                  data: {
                    attributes: {
                      quizAndTipsBlock: {
                        data: {
                          id: "1",
                          attributes: {
                            title: "Titre modifi√©",
                            displayBlock: false,
                            displayQuiz: false,
                            quiz: {
                              data: {
                                id: "2",
                                attributes: {
                                  title: "Nom du quiz",
                                  publishedAt: "2022-11-20T02:34:14.535Z",
                                },
                              },
                            },
                            displayTips: false,
                            tips: {
                              data: [],
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          ],
        },
        services: {
          data: [
            {
              attributes: {
                serviceInstance: [
                  {
                    __typename: "ComponentMsdEditorial",
                    id: "4",
                    editorialServices: {
                      data: [
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoQuizzesSubService",
                                isActivated: true,
                                quizzes: {
                                  data: [
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Nom du quiz",
                                        publishedAt: "2022-11-20T02:34:14.535Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "truc",
                                        publishedAt: "2022-11-18T10:16:08.936Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoTipsSubService",
                                isActivated: true,
                                tips: {
                                  data: [
                                    {
                                      id: "5",
                                      attributes: {
                                        title: "nah",
                                        publishedAt: "2022-11-19T00:14:37.403Z",
                                      },
                                    },
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Titre de l'astuce 1",
                                        publishedAt: "2022-11-19T00:14:08.232Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "Titre de l'astuce 2",
                                        publishedAt: "2022-11-18T10:17:11.158Z",
                                      },
                                    },
                                    {
                                      id: "3",
                                      attributes: {
                                        title: "Titre de l'astuce 3",
                                        publishedAt: "2022-11-19T00:14:20.490Z",
                                      },
                                    },
                                    {
                                      id: "4",
                                      attributes: {
                                        title: "Titre de l'astuce 4",
                                        publishedAt: "2022-11-19T00:14:30.273Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    error: new Error("An error occured"),
  },
];
