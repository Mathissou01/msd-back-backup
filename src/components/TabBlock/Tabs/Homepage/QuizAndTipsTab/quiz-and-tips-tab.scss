.c-QuizAndTipsTab {
  position: relative;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  width: 100%;
  height: 100%;

  &__Title {
    margin-bottom: 18px;
    font-size: 28px;
  }

  &__Form {
    display: flex;
    flex-direction: column;
    gap: 24px;
    width: 100%;
  }

  &__Group {
    display: flex;
    flex-direction: column;
    gap: 10px;

    &_short {
      width: 60%;
    }
  }

  &__Buttons {
    display: flex;
    flex-direction: row;
    gap: 16px;
  }
}
