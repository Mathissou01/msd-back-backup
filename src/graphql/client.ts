import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache({
      typePolicies: {
        UploadFolder: {
          merge: true,
        },
        UploadFile: {
          merge: true,
        },
        Contract: {
          merge: true,
        },
        ContractCustomization: {
          merge: true,
        },
        New: {
          merge: true,
        },
        Tip: {
          merge: true,
        },
        RecyclingGuideService: {
          merge: true,
        },
        WasteForm: {
          merge: true,
        },
        WasteFamily: {
          merge: true,
        },
        Flow: {
          merge: true,
        },
        FlowColor: {
          merge: true,
        },
        DropOffMap: {
          merge: true,
        },
        Request: {
          merge: true,
        },
        RequestAggregate: {
          merge: true,
        },
      },
    }),
  });
}

export default client;
