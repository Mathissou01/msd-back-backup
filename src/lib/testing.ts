import { prettyDOM } from "@testing-library/react";

export function removeUnstableHtmlPropertiesForTesting(
  htmlElement: HTMLElement,
  attributeName: string,
) {
  const domHTML = prettyDOM(htmlElement, Infinity).toString();
  if (!domHTML) return undefined;
  const regex = new RegExp(`${attributeName}(.*)"(.*)"`, "g");
  return domHTML.replaceAll(regex, "");
}
