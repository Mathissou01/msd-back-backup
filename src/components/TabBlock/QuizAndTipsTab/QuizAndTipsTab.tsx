import { format, parseJSON } from "date-fns";
import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, {
  ReactNode,
  useCallback,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  GetQuizAndTipsBlockTabDocument,
  QuizEntity,
  TipEntity,
  useGetQuizAndTipsBlockTabQuery,
  useUpdateQuizAndTipsBlockTabMutation,
} from "../../../graphql/codegen/generated-types";
import { FocusFirstElement, removeNulls } from "../../../lib/utilities";
import { extractQuizAndTipsBlock } from "../../../lib/graphql-data";
import { useContract } from "../../../hooks/useContract";
import CommonButton from "../../Common/CommonButton/CommonButton";
import FormInput from "../../Form/FormInput/FormInput";
import FormCheckbox from "../../Form/FormCheckbox/FormCheckbox";
import FormModalInput from "../../Form/FormModalInput/FormModalInput";
import FormSelect from "../../Form/FormSelect/FormSelect";
import FormMultiselect, {
  IOptionWrapper,
} from "../../Form/FormMultiselect/FormMultiselect";
import CommonSpinner from "../../Common/CommonSpinner/CommonSpinner";
import "./quiz-and-tips-tab.scss";

interface IQuizAndTipsBlock {
  id: string;
  titleContent: string;
  displayBlock: boolean;
  displayQuiz: boolean;
  quiz?: QuizEntity | null;
  displayTips: boolean;
  tips?: Array<TipEntity> | null;
}

export default function QuizAndTipsTab() {
  /* Static Data */
  const formLabels = {
    title: "Quiz & Astuces",
    displayBlock: "Afficher ce bloc",
    titleContent: "Titre du bloc",
    displayQuiz: "Afficher un quiz",
    quiz: "Quiz à afficher",
    quizButton: "Choisir le quiz",
    quizModal: "Sélection du quiz",
    quizModalType: "Quiz",
    displayTips: "Afficher des astuces",
    tips: "Astuces à afficher",
    tipsButton: "Sélectionner les astuces à afficher",
    tipsModal: "Sélection des astuces",
    tipsModalType: "Astuce",
  };
  const submitButtonLabel = "Enregistrer les modifications";
  const cancelButtonLabel = "Annuler les modifications";

  /* Methods */
  function quizDisplayTransformFunction(
    quiz: QuizEntity | undefined,
  ): ReactNode {
    return <p>{quiz?.attributes?.title ?? ""}</p>;
  }

  function quizSelectDisplayTransformFunction(quiz: QuizEntity): string {
    return quiz.attributes?.title ?? "";
  }

  function onQuizModalSubmit(submitData: {
    [key: string]: QuizEntity | undefined;
  }) {
    const quiz = Object.values(submitData)[0];
    setValue("quiz", quiz, { shouldDirty: true });
  }

  function tipsDisplayTransformFunction(
    tips: Array<TipEntity | undefined>,
  ): ReactNode {
    return tips?.map((tip: TipEntity | undefined, index) => {
      if (tip && tip.id) {
        return (
          <p key={tip.id + index}>
            {`${tip.attributes?.title} - ${format(
              parseJSON(tip.attributes?.publishedAt),
              "dd/mm/yyyy",
            )}` ?? ""}
          </p>
        );
      }
    });
  }

  function tipSelectDisplayTransformFunction(tip: TipEntity): string {
    return (
      `${tip.attributes?.title} - ${format(
        parseJSON(tip.attributes?.publishedAt),
        "dd/mm/yyyy",
      )}` ?? ""
    );
  }

  function onTipsModalSubmit(submitData: {
    [key: string]: Array<TipEntity | undefined>;
  }) {
    const tips = Object.values(submitData)?.filter(removeNulls);
    setValue("tips", tips, { shouldDirty: true });
  }

  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        quizAndTipsBlockId: submitData["id"],
        data: {
          titleContent: submitData["titleContent"],
          displayBlock: submitData["displayBlock"],
          displayQuiz: submitData["displayQuiz"],
          quiz: submitData["quiz"]?.id ?? null,
          displayTips: submitData["displayTips"],
          tips:
            submitData["tips"]?.map((tip: TipEntity) => tip?.id ?? null) ??
            null,
        },
      };
      await updateQuizAndTipsBlock({
        variables,
        refetchQueries: [
          {
            query: GetQuizAndTipsBlockTabDocument,
            variables: { contractId },
          },
          "getQuizAndTipsBlockTab",
        ],
      });
      return new Promise<void>((resolve) => {
        setTimeout(() => resolve(), 1000);
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const { loading, error, data } = useGetQuizAndTipsBlockTabQuery({
    variables: { contractId },
  });
  const [
    updateQuizAndTipsBlock,
    { loading: mutationLoading, error: mutationError },
  ] = useUpdateQuizAndTipsBlockTabMutation();

  /* Local Data */
  const [isShowingSpinner, setIsShowingSpinner] = useState(false);
  const [quizAndTipsData, setQuizAndTipsData] = useState<IQuizAndTipsBlock>();
  const [quizzesData, setQuizzesData] = useState<Array<QuizEntity>>([]);
  const [tipsData, setTipsData] = useState<Array<IOptionWrapper<TipEntity>>>(
    [],
  );
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, setValue, watch, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (
      data &&
      data.contractCustomizations?.data &&
      (data.quizSubServices?.data || data?.tipSubServices?.data)
    ) {
      const { quizAndTipsBlock, quizzes, tips } = extractQuizAndTipsBlock(data);
      if (quizAndTipsBlock?.id && quizAndTipsBlock?.attributes) {
        const mappedData: IQuizAndTipsBlock = {
          id: quizAndTipsBlock.id,
          titleContent:
            quizAndTipsBlock.attributes.titleContent ?? "Quiz & Astuces",
          displayBlock: quizAndTipsBlock.attributes.displayBlock,
          displayQuiz: quizAndTipsBlock.attributes.displayQuiz,
          quiz: quizAndTipsBlock.attributes.quiz?.data,
          displayTips: quizAndTipsBlock.attributes.displayTips,
          tips: quizAndTipsBlock?.attributes.tips?.data,
        };
        setQuizAndTipsData(mappedData);
        form.reset(mappedData);
      }
      if (quizzes && quizzes?.length > 0) {
        setQuizzesData(quizzes);
      }
      if (tips && tips?.length > 0) {
        const mappedOptions: Array<IOptionWrapper<TipEntity> | null> = tips.map(
          (tip) => {
            return tip ? { option: tip } : null;
          },
        );
        setTipsData(
          mappedOptions?.filter(
            (e): e is Exclude<typeof e, null> => e !== null,
          ),
        );
      }
    }
  }, [form, data]);

  const spinnerTimerRef = useRef<NodeJS.Timeout>();
  useEffect(() => {
    if (isSubmitting) {
      if (!isShowingSpinner) {
        spinnerTimerRef.current = setTimeout(() => {
          setIsShowingSpinner(true);
        }, 200);
      }
    } else {
      clearTimeout(spinnerTimerRef.current);
      setIsShowingSpinner(false);
    }
  }, [isShowingSpinner, isSubmitting]);

  const focusRef = useCallback((node: HTMLFormElement) => {
    FocusFirstElement(node);
  }, []);

  {
    // TODO: layout shift, handle error redirect,
  }
  if (loading) return <CommonSpinner />;
  if (error) return <span>{error?.message}</span>;
  if (mutationError) return <span>{mutationError?.message}</span>;

  return (
    <div className="c-QuizAndTipsTab">
      {isShowingSpinner && <CommonSpinner isCover={true} />}
      <h2 className="c-QuizAndTipsTab__Title">{formLabels.title}</h2>
      <FormProvider {...form}>
        <form
          className="c-QuizAndTipsTab__Form"
          onSubmit={handleSubmit(onSubmitValid)}
          ref={focusRef}
        >
          <div className="c-QuizAndTipsTab__Group c-QuizAndTipsTab__Group_short">
            <FormCheckbox name="displayBlock" label={formLabels.displayBlock} />
            <FormInput
              type="text"
              name="titleContent"
              label={formLabels.titleContent}
              isRequired={true}
              isDisabled={mutationLoading}
              defaultValue={quizAndTipsData?.titleContent}
            />
          </div>
          <div className="c-QuizAndTipsTab__Group">
            <FormCheckbox name="displayQuiz" label={formLabels.displayQuiz} />
            <FormModalInput<QuizEntity>
              name="quiz"
              label={formLabels.quiz}
              displayTransform={quizDisplayTransformFunction}
              buttonLabel={formLabels.quizButton}
              modalTitle={formLabels.quizModal}
              onSubmit={onQuizModalSubmit}
              formValidationMode={formValidationMode}
              isDisabled={mutationLoading}
            >
              <FormSelect<QuizEntity>
                name="quizSelect"
                label={formLabels.quizModalType}
                displayTransform={quizSelectDisplayTransformFunction}
                options={quizzesData}
                optionKey={"id"}
                defaultValue={watch("quiz")}
              />
            </FormModalInput>
          </div>
          <div className="c-QuizAndTipsTab__Group">
            <FormCheckbox name="displayTips" label={formLabels.displayTips} />
            <FormModalInput<Array<TipEntity>>
              name="tips"
              label={formLabels.tips}
              displayTransform={tipsDisplayTransformFunction}
              buttonLabel={formLabels.tipsButton}
              modalTitle={formLabels.tipsModal}
              onSubmit={onTipsModalSubmit}
              formValidationMode={formValidationMode}
              isDisabled={mutationLoading}
            >
              <FormMultiselect<TipEntity>
                name="tipsSelect"
                label={formLabels.tipsModalType}
                displayTransform={tipSelectDisplayTransformFunction}
                options={tipsData}
                selectAmount={3}
                optionKey={"id"}
                defaultValues={watch("tips")}
              />
            </FormModalInput>
          </div>
          <div className="c-QuizAndTipsTab__Buttons">
            <CommonButton
              type="submit"
              label={submitButtonLabel}
              style="primary"
              isDisabled={!isDirty}
            />
            <CommonButton
              type="button"
              label={cancelButtonLabel}
              onClick={onCancel}
              isDisabled={!isDirty}
            />
          </div>
        </form>
      </FormProvider>
    </div>
  );
}
