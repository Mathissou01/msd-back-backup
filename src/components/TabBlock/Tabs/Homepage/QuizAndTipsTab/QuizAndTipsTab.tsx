import { parseJSON } from "date-fns";
import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, { ReactNode, useEffect, useState } from "react";
import {
  AudienceEntity,
  Enum_Tip_Status,
  QuizEntity,
  TipEntity,
  useGetQuizAndTipsBlockTabQuery,
  useUpdateQuizAndTipsBlockTabMutation,
} from "../../../../../graphql/codegen/generated-types";
import { formatDate, removeNulls } from "../../../../../lib/utilities";
import { extractQuizAndTipsBlock } from "../../../../../lib/graphql-data";
import { useFocusFirstElement } from "../../../../../hooks/useFocusFirstElement";
import { useContract } from "../../../../../hooks/useContract";
import CommonLoader from "../../../../Common/CommonLoader/CommonLoader";
import CommonButton from "../../../../Common/CommonButton/CommonButton";
import FormInput from "../../../../Form/FormInput/FormInput";
import FormCheckbox from "../../../../Form/FormCheckbox/FormCheckbox";
import FormModalButtonInput from "../../../../Form/FormModalButtonInput/FormModalButtonInput";
import FormSelect from "../../../../Form/FormSelect/FormSelect";
import FormMultiselect, {
  IOptionWrapper,
} from "../../../../Form/FormMultiselect/FormMultiselect";
import "./quiz-and-tips-tab.scss";

interface IQuizAndTipsBlock {
  id: string;
  titleContent: string;
  displayBlock: boolean;
  displayQuiz: boolean;
  quiz?: QuizEntity | null;
  displayTips: boolean;
  tips?: Array<TipEntity> | null;
}

interface IQuizModalFields {
  quizSelect: QuizEntity;
}

interface ITipsModalFields {
  tipsSelect_0: TipEntity;
  tipsSelect_1: TipEntity;
  tipsSelect_2: TipEntity;
}

interface IQuizAndTipsTabProps {
  audience?: AudienceEntity;
}

export default function QuizAndTipsTab({ audience }: IQuizAndTipsTabProps) {
  /* Static Data */
  const formLabels = {
    title: "Quiz & Astuces",
    displayBlock: "Afficher ce bloc",
    titleContent: "Titre du bloc",
    displayQuiz: "Afficher un quiz",
    quiz: "Quiz à afficher",
    quizButton: "Choisir le quiz",
    quizModal: "Sélection du quiz",
    quizModalType: "Quiz",
    displayTips: "Afficher des astuces",
    tips: "Astuces à afficher",
    tipsButton: "Sélectionner les astuces à afficher",
    tipsModal: "Sélection des astuces",
    tipsModalType: "Astuce",
  };
  const submitButtonLabel = "Enregistrer les modifications";
  const cancelButtonLabel = "Annuler les modifications";
  const status = Enum_Tip_Status.Published;

  /* Methods */
  function quizDisplayTransformFunction(
    quiz?: QuizEntity | undefined,
  ): ReactNode {
    return <p>{quiz?.attributes?.title ?? ""}</p>;
  }

  function quizSelectDisplayTransformFunction(quiz: QuizEntity): string {
    return quiz.attributes?.title ?? "";
  }

  function onQuizModalSubmit(submitData: IQuizModalFields) {
    return Object.values(submitData)[0];
  }

  function tipsDisplayTransformFunction(
    tips?: Array<TipEntity | undefined>,
  ): ReactNode {
    return tips?.map((tip: TipEntity | undefined, index) => {
      if (tip && tip.id) {
        return (
          <p key={tip.id + index}>
            {`${tip.attributes?.title} - ${formatDate(
              parseJSON(tip.attributes?.publishedDate),
              "dd/MM/yyyy",
            )}`}
          </p>
        );
      }
    });
  }

  function tipSelectDisplayTransformFunction(tip: TipEntity): string {
    return `${tip.attributes?.title} - ${formatDate(
      parseJSON(tip.attributes?.publishedDate),
      "dd/MM/yyyy",
    )}`;
  }

  function onTipsModalSubmit(submitData: ITipsModalFields) {
    return [
      submitData.tipsSelect_0,
      submitData.tipsSelect_1,
      submitData.tipsSelect_2,
    ];
  }

  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        quizAndTipsBlockId: submitData["id"],
        data: {
          titleContent: submitData["titleContent"],
          displayBlock: submitData["displayBlock"],
          displayQuiz: submitData["displayQuiz"],
          quiz: submitData["quiz"]?.id ?? null,
          displayTips: submitData["displayTips"],
          tips:
            submitData["tips"]
              ?.map((tip: TipEntity) => tip?.id ?? null)
              .filter(removeNulls) ?? null,
        },
      };
      return updateQuizAndTipsBlock({
        variables,
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const getQuizAndTipsBlockTabVariables: {
    contractId: string;
    status: Enum_Tip_Status;
    audienceId: string;
  } = {
    contractId,
    status,
    audienceId: "",
  };
  if (audience && audience.id) {
    getQuizAndTipsBlockTabVariables.audienceId = audience.id;
  }
  const { loading, error, data } = useGetQuizAndTipsBlockTabQuery({
    variables: getQuizAndTipsBlockTabVariables,
    fetchPolicy: "network-only",
  });
  const [
    updateQuizAndTipsBlock,
    { loading: mutationLoading, error: mutationError },
  ] = useUpdateQuizAndTipsBlockTabMutation({
    refetchQueries: ["getQuizAndTipsBlockTab"],
    awaitRefetchQueries: true,
  });

  /* Local Data */
  const [quizAndTipsData, setQuizAndTipsData] = useState<IQuizAndTipsBlock>();
  const [quizzesData, setQuizzesData] = useState<
    Array<IOptionWrapper<QuizEntity>>
  >([]);
  const [tipsData, setTipsData] = useState<Array<IOptionWrapper<TipEntity>>>(
    [],
  );
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, watch, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (
      data &&
      data.contractCustomizations?.data &&
      (data.quizSubServices?.data || data?.tipSubServices?.data)
    ) {
      const { quizAndTipsBlock, quizzes, tips } = extractQuizAndTipsBlock(data);
      if (quizAndTipsBlock?.id && quizAndTipsBlock?.attributes) {
        const mappedData: IQuizAndTipsBlock = {
          id: quizAndTipsBlock.id,
          titleContent:
            quizAndTipsBlock.attributes.titleContent ?? "Quiz & Astuces",
          displayBlock: quizAndTipsBlock.attributes.displayBlock,
          displayQuiz: quizAndTipsBlock.attributes.displayQuiz,
          quiz: quizAndTipsBlock.attributes.quiz?.data,
          displayTips: quizAndTipsBlock.attributes.displayTips,
          tips: quizAndTipsBlock?.attributes.tips?.data,
        };
        setQuizAndTipsData(mappedData);
        form.reset(mappedData);
      }
      if (quizzes && quizzes?.length > 0) {
        const mappedOptions: Array<IOptionWrapper<QuizEntity> | null> =
          quizzes.map((quiz) => {
            return quiz ? { option: quiz } : null;
          });
        setQuizzesData(mappedOptions?.filter(removeNulls));
      }
      if (tips && tips?.length > 0) {
        const mappedOptions: Array<IOptionWrapper<TipEntity> | null> = tips.map(
          (tip) => {
            return tip ? { option: tip } : null;
          },
        );
        setTipsData(mappedOptions?.filter(removeNulls));
      }
    }
  }, [form, data]);

  return (
    <div
      className="c-QuizAndTipsTab"
      id="panel-quizAndTips"
      role="tabpanel"
      aria-labelledby="tab-quizAndTips"
    >
      <CommonLoader
        isLoading={loading || isSubmitting || mutationLoading}
        isShowingContent={isSubmitting || mutationLoading}
        hasDelay={isSubmitting || mutationLoading}
        errors={[error, mutationError]}
      >
        <h2 className="c-QuizAndTipsTab__Title">{formLabels.title}</h2>
        <FormProvider {...form}>
          <form
            className="c-QuizAndTipsTab__Form"
            onSubmit={handleSubmit(onSubmitValid)}
            ref={useFocusFirstElement()}
          >
            <div className="c-QuizAndTipsTab__Group c-QuizAndTipsTab__Group_short">
              <FormCheckbox
                name="displayBlock"
                label={formLabels.displayBlock}
              />
              <FormInput
                type="text"
                name="titleContent"
                label={formLabels.titleContent}
                isRequired={true}
                isDisabled={mutationLoading}
                defaultValue={quizAndTipsData?.titleContent}
              />
            </div>
            <div className="c-QuizAndTipsTab__Group">
              <FormCheckbox name="displayQuiz" label={formLabels.displayQuiz} />
              <FormModalButtonInput<QuizEntity, IQuizModalFields>
                name="quiz"
                label={formLabels.quiz}
                displayTransform={quizDisplayTransformFunction}
                buttonLabel={formLabels.quizButton}
                modalTitle={formLabels.quizModal}
                onModalSubmit={onQuizModalSubmit}
                isDisabled={mutationLoading}
              >
                <FormSelect<QuizEntity>
                  name="quizSelect"
                  label={formLabels.quizModalType}
                  displayTransform={quizSelectDisplayTransformFunction}
                  options={quizzesData}
                  optionKey={"id"}
                  defaultValue={watch("quiz")}
                />
              </FormModalButtonInput>
            </div>
            <div className="c-QuizAndTipsTab__Group">
              <FormCheckbox name="displayTips" label={formLabels.displayTips} />
              <FormModalButtonInput<Array<TipEntity>, ITipsModalFields>
                name="tips"
                label={formLabels.tips}
                displayTransform={tipsDisplayTransformFunction}
                buttonLabel={formLabels.tipsButton}
                modalTitle={formLabels.tipsModal}
                onModalSubmit={onTipsModalSubmit}
                isDisabled={mutationLoading}
              >
                <FormMultiselect<TipEntity>
                  name="tipsSelect"
                  label={formLabels.tipsModalType}
                  displayTransform={tipSelectDisplayTransformFunction}
                  options={tipsData}
                  selectAmount={3}
                  optionKey={"id"}
                  defaultValues={watch("tips")}
                />
              </FormModalButtonInput>
            </div>
            <div className="c-QuizAndTipsTab__Buttons">
              <CommonButton
                type="submit"
                label={submitButtonLabel}
                style="primary"
                picto="check"
                isDisabled={!isDirty}
              />
              <CommonButton
                type="button"
                label={cancelButtonLabel}
                picto="cross"
                onClick={onCancel}
                isDisabled={!isDirty}
              />
            </div>
          </form>
        </FormProvider>
      </CommonLoader>
    </div>
  );
}
