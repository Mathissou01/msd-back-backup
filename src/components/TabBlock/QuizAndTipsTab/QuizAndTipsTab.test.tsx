import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import React from "react";
import { defaultMockData } from "../../../../__mocks__/quizAndTipsTabMockData";
import QuizAndTipsTab from "./QuizAndTipsTab";

describe("QuizAndTipsTab", () => {
  it("renders loading and loaded state", async () => {
    const { container } = render(
      <MockedProvider mocks={defaultMockData} addTypename={false}>
        <QuizAndTipsTab />
      </MockedProvider>,
    );

    expect(await screen.findByTestId("common-spinner")).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
    expect(await screen.findByText("Nom du quiz")).toBeInTheDocument();
    expect(
      await screen.findByDisplayValue("Titre modifiÃ©"),
    ).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
  });

  // it("should show error UI", async () => {
  //   const { container } = render(
  //     <MockedProvider mocks={errorMockData} addTypename={false}>
  //       <QuizAndTipsTab />
  //     </MockedProvider>,
  //   );
  //
  //   expect(await screen.findByText("Loading...")).toBeInTheDocument();
  //   expect(await screen.findByText("An error occurred")).toBeInTheDocument();
  //   expect(await container).toMatchSnapshot();
  // });
});
