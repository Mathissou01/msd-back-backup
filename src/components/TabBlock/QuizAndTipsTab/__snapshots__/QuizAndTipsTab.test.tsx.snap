// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizAndTipsTab renders loading and loaded state 1`] = `
<div>
  <div
    class="c-CommonSpinner__Wrapper "
  >
    <div
      class="c-CommonSpinner"
      data-testid="common-spinner"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;

exports[`QuizAndTipsTab renders loading and loaded state 2`] = `
<div>
  <div
    class="c-QuizAndTipsTab"
  >
    <h2
      class="c-QuizAndTipsTab__Title"
    >
      Quiz & Astuces
    </h2>
    <form
      class="c-QuizAndTipsTab__Form"
    >
      <div
        class="c-QuizAndTipsTab__Group c-QuizAndTipsTab__Group_short"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayBlock"
            name="displayBlock"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayBlock"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher ce bloc 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormInput"
        >
          <label
            class="c-FormLabel"
            for="titleContent"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Titre du bloc *
              </span>
            </div>
          </label>
          <input
            aria-errormessage="titleContent_error"
            aria-invalid="false"
            class="c-FormInput__Input "
            data-testid="form-input"
            id="titleContent"
            name="titleContent"
            type="text"
            value="Titre modifié"
          />
        </div>
      </div>
      <div
        class="c-QuizAndTipsTab__Group"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayQuiz"
            name="displayQuiz"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayQuiz"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher un quiz 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormModalInput"
          data-testid="form-modal-input"
        >
          <label
            class="c-FormLabel"
            for="quiz_button"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Quiz à afficher 
              </span>
            </div>
          </label>
          <input
            class="c-FormModalInput__Input"
            disabled=""
            name="quiz"
            type="hidden"
            value="[object Object]"
          />
          <p>
            Nom du quiz
          </p>
          <div
            aria-errormessage="quiz_error"
            aria-invalid="false"
          >
            <button
              class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
              id="quiz_button"
              type="button"
            >
              Choisir le quiz
            </button>
          </div>
        </div>
      </div>
      <div
        class="c-QuizAndTipsTab__Group"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayTips"
            name="displayTips"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayTips"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher des astuces 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormModalInput"
          data-testid="form-modal-input"
        >
          <label
            class="c-FormLabel"
            for="tips_button"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Astuces à afficher 
              </span>
            </div>
          </label>
          <input
            class="c-FormModalInput__Input"
            disabled=""
            name="tips"
            type="hidden"
            value=""
          />
          <div
            aria-errormessage="tips_error"
            aria-invalid="false"
          >
            <button
              class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
              id="tips_button"
              type="button"
            >
              Sélectionner les astuces à afficher
            </button>
          </div>
        </div>
      </div>
      <div
        class="c-QuizAndTipsTab__Buttons"
      >
        <button
          class="c-CommonButton c-CommonButton_primary c-CommonButton_fontSmall"
          disabled=""
          type="submit"
        >
          Enregistrer les modifications
        </button>
        <button
          class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
          disabled=""
          type="button"
        >
          Annuler les modifications
        </button>
      </div>
    </form>
  </div>
</div>
`;
